hw/mcu/stm/stm32_common/src/hal_os_tick.c

```diff
@@ -76,3 +79,111 @@ os_tick_init(uint32_t os_ticks_per_sec, int prio)
     DBGMCU->CR |= (DBGMCU_CR_DBG_STOP | DBGMCU_CR_DBG_STANDBY);
 #endif
 }
+
+#else
+
+struct hal_os_tick {
+    uint32_t rtc_cnt;
+};
+
+struct hal_os_tick g_hal_os_tick;
+
+static void
+stm32_os_tick_update_rtc(void)
+{
+    int ticks;
+    uint32_t rtc_cnt;
+
+    /* Clear all flags. */
+    RTC->CRL = 0;
+    /* Read current value of counter. */
+    rtc_cnt = (RTC->CNTH << 16u) | RTC->CNTL;
+    if (RTC->CRL & RTC_CRL_SECF_Msk) {
+        /* Read again in case it was just incremented. */
+        rtc_cnt = (RTC->CNTH << 16u) | RTC->CNTL;
+    }
+    ticks = rtc_cnt - g_hal_os_tick.rtc_cnt;
+    g_hal_os_tick.rtc_cnt = rtc_cnt;
+
+    os_time_advance(ticks);
+}
+
+static void
+stm32_os_tick_rtc_handler(void)
+{
+    os_trace_isr_enter();
+    stm32_os_tick_update_rtc();
+    os_trace_isr_exit();
+}
+
+void
+os_tick_idle(os_time_t ticks)
+{
+    uint32_t rtc_cnt;
+    OS_ASSERT_CRITICAL();
+
+    if (ticks > 1) {
+        while ((RTC->CRL & RTC_CRL_RTOFF_Msk) == 0);
+        RTC->CRL = RTC_CRL_CNF_Msk;
+        rtc_cnt = (RTC->CNTH << 16u) + RTC->CNTL + ticks;
+        if (RTC->CRL & RTC_CRL_SECF_Msk) {
+            /* Just ticked reduce ticks by one. */
+            rtc_cnt = (RTC->CNTH << 16u) + RTC->CNTL + ticks - 1;
+        }
+        /* Set alarm. */
+        RTC->ALRH = rtc_cnt >> 16u;
+        RTC->ALRL = (uint16_t)rtc_cnt;
+        /* Only alarm interrupt. */
+        RTC->CRH = RTC_CRH_ALRIE_Msk;
+        /* Finish configuration. */
+        RTC->CRL &= ~RTC_CRL_CNF_Msk;
+    } else {
+        /* Disable alarm, enable tick interrupt. */
+        RTC->CRH = RTC_CRH_SECIE_Msk;
+    }
+
+    __DSB();
+    __WFI();
+
+    if (ticks > 0) {
+        /*
+         * Update OS time before anything else when coming out of
+         * the tickless regime.
+         */
+        stm32_os_tick_update_rtc();
+    }
+}
+
+void
+os_tick_init(uint32_t os_ticks_per_sec, int prio)
+{
+    uint32_t sr;
+
+    __HAL_RCC_RTC_CONFIG(RCC_RTCCLKSOURCE_LSE);
+    __HAL_RCC_RTC_ENABLE();
```

Comment:
This may not be enough. ST HAL code for changing clock source does a lot of other things since changing it may not be possible after it was first set (for example if bootloader was not configured to use LSE and application is change may not be applied).
Have you checked that it will work in this scenario?
Maybe ST HAL could be used for this?